File: architecture/modules/storage/node/src/cloud.ts
import {Readable,PassThrough} from 'node:stream';
import {pipeline} from 'node:stream/promises';
import {StorageError,call,validKey,validSize,validContentType,pageLimit,ttl} from './contracts.ts';
import type {StorageProvider,ObjectInfo,SignedRequest,PutInput,ListInput,ProviderName} from './contracts.ts';
import {readable,sizeGuard} from './streams.ts';
export interface CloudCredentials {accessKeyId:string;secretAccessKey:string;sessionToken?:string;expiration?:Date}
export interface CloudOptions {bucket:string;region:string;endpoint?:string;publicEndpoint?:string;credentials?:()=>Promise<CloudCredentials>;allowHTTP?:boolean;forcePathStyle?:boolean}
export function endpoint(value:string|undefined,allowHTTP=false):string|undefined {
  if(!value)return undefined;const u=new URL(value);
  if((u.protocol!=='https:'&&!(allowHTTP&&u.protocol==='http:'&&['127.0.0.1','localhost','[::1]'].includes(u.hostname)))||u.username||u.password||u.search||u.hash||u.pathname!=='/')throw new StorageError('CONFIGURATION','Invalid storage endpoint');
  return u.origin;
}
export function cloudOptions(o:CloudOptions):CloudOptions {
  if(!o.bucket||!/^[-a-z0-9.]{3,100}$/.test(o.bucket)||!/^[-a-z0-9]{2,64}$/.test(o.region))throw new StorageError('CONFIGURATION','Bucket and region are required');
  endpoint(o.endpoint,o.allowHTTP);endpoint(o.publicEndpoint,o.allowHTTP);
  const internal=(value:string)=>/internal|privatelink|vpce[.-]|localhost|^127\.|^10\.|^192\.168\.|^172\.(?:1[6-9]|2[0-9]|3[01])\./.test(new URL(value).hostname);
  if(!o.allowHTTP&&((o.endpoint&&internal(o.endpoint)&&!o.publicEndpoint)||(o.publicEndpoint&&internal(o.publicEndpoint))))throw new StorageError('CONFIGURATION','A public storage endpoint is required for browser signatures');return o;
}
export function expiresIn(seconds:number,credentials?:CloudCredentials):number{
  ttl(seconds);if(!credentials?.expiration)return seconds;const remaining=Math.floor((credentials.expiration.getTime()-Date.now())/1000)-10;
  if(!Number.isFinite(remaining)||remaining<1)throw new StorageError('CONFIGURATION','Storage credentials expired');return Math.min(seconds,remaining);
}
export function signed(method:'PUT'|'GET',url:string,headers:Record<string,string>,seconds:number):SignedRequest{return {method,url,headers,expiresAt:new Date(Date.now()+seconds*1000).toISOString()};}
export function fromHeaders(key:string,h:Headers):ObjectInfo {
  const raw=h.get('content-length'),size=raw===null?NaN:Number(raw);if(!Number.isSafeInteger(size)||size<0)throw new StorageError('UNAVAILABLE','Storage response has no valid content length');
  return {key,size,contentType:h.get('content-type')||'application/octet-stream',etag:h.get('etag')||undefined,modifiedAt:h.get('last-modified')||undefined};
}
async function responseOK(r:Response){if(!r.ok){await r.body?.cancel();throw {status:r.status};}}
export function signedProvider(name:ProviderName,operations:{
  sign:(method:'PUT'|'GET',key:string,input:{contentType?:string;size?:number;expiresIn:number},publicURL:boolean)=>Promise<SignedRequest>;
  head:StorageProvider['head'];delete:StorageProvider['delete'];list:StorageProvider['list'];
}):StorageProvider{
  const provider:StorageProvider={
    name,capabilities:{directUpload:true,multipart:false,resumable:false,extensions:[]},
    signUpload:(key,input)=>call(async()=>{validKey(key);validSize(input.size);validContentType(input.contentType);ttl(input.expiresIn);return operations.sign('PUT',key,input,true);}),
    signDownload:(key,input)=>call(async()=>{validKey(key);ttl(input.expiresIn);return operations.sign('GET',key,input,true);}),
    head:(key,opts)=>call(async()=>{validKey(key);opts?.signal?.throwIfAborted();return operations.head(key,opts);}),
    delete:(key,opts)=>call(async()=>{validKey(key);opts?.signal?.throwIfAborted();try{await operations.delete(key,opts);}catch(e){if((e as {status?:number;statusCode?:number}).status!==404&&(e as {statusCode?:number}).statusCode!==404)throw e;}}),
    list:(input={})=>call(async()=>{pageLimit(input.limit);if(input.prefix&&input.prefix.length>1024)throw new StorageError('INVALID_INPUT');if(input.cursor&&input.cursor.length>8192)throw new StorageError('INVALID_INPUT');input.signal?.throwIfAborted();return operations.list(input);}),
    async get(key,{signal}={}){return call(async()=>{validKey(key);const auth=await operations.sign('GET',key,{expiresIn:300},false);const response=await fetch(auth.url,{headers:auth.headers,signal:AbortSignal.any([AbortSignal.timeout(300000),...(signal?[signal]:[])])});await responseOK(response);if(!response.body)throw new StorageError('UNAVAILABLE');try{return {info:fromHeaders(key,response.headers),body:Readable.fromWeb(response.body as import('node:stream/web').ReadableStream)};}catch(error){await response.body.cancel();throw error;}});},
    async put(key,input:PutInput){return call(async()=>{
      validKey(key);validSize(input.size);validContentType(input.contentType);const auth=await operations.sign('PUT',key,{size:input.size,contentType:input.contentType,expiresIn:300},false);
      const guard=sizeGuard(input.size),bridge=new PassThrough(),signal=AbortSignal.any([AbortSignal.timeout(300000),...(input.signal?[input.signal]:[])]);
      const copying=pipeline(readable(input.body),guard.stream,bridge,{signal});
      const headers={...auth.headers,'content-type':input.contentType,'content-length':String(input.size)};
      const sending=(async()=>{const response=await fetch(auth.url,{method:'PUT',headers,body:bridge as unknown as BodyInit,duplex:'half',signal} as RequestInit);await responseOK(response);await response.body?.cancel();return response;})();
      try{const [,response]=await Promise.all([copying,sending]);return {key,size:input.size,contentType:input.contentType,etag:response.headers.get('etag')||undefined};}catch(e){bridge.destroy();await Promise.allSettled([copying,sending]);throw e;}
    });}
  };return provider;
}

File: architecture/modules/storage/go/cloud.go
package storage

import (
	"context"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"
)

type Credentials struct {
	AccessKeyID, SecretAccessKey, SessionToken string
	Expiration                                 time.Time
}
type CloudOptions struct {
	Bucket, Region, Endpoint, PublicEndpoint string
	AllowHTTP, ForcePathStyle                bool
	Credentials                              func(context.Context) (Credentials, error)
}

func (o CloudOptions) Validate() error {
	if o.Bucket == "" || o.Region == "" {
		return fail("CONFIGURATION")
	}
	for _, e := range []string{o.Endpoint, o.PublicEndpoint} {
		if e == "" {
			continue
		}
		u, err := url.Parse(e)
		if err != nil || u.User != nil || u.RawQuery != "" || u.Fragment != "" || (u.Path != "" && u.Path != "/") || u.Host == "" {
			return fail("CONFIGURATION")
		}
		if u.Scheme != "https" && !(o.AllowHTTP && u.Scheme == "http" && (u.Hostname() == "127.0.0.1" || u.Hostname() == "localhost")) {
			return fail("CONFIGURATION")
		}
	}
	private := regexp.MustCompile(`internal|privatelink|vpce[.-]|localhost|^127\.|^10\.|^192\.168\.|^172\.(?:1[6-9]|2[0-9]|3[01])\.`)
	internal := func(endpoint string) bool {
		u, _ := url.Parse(endpoint)
		return u != nil && private.MatchString(u.Hostname())
	}
	if !o.AllowHTTP && ((o.Endpoint != "" && internal(o.Endpoint) && o.PublicEndpoint == "") || (o.PublicEndpoint != "" && internal(o.PublicEndpoint))) {
		return fail("CONFIGURATION")
	}
	return nil
}
func (o CloudOptions) endpoint(public bool) string {
	if public && o.PublicEndpoint != "" {
		return o.PublicEndpoint
	}
	return o.Endpoint
}
func duration(ttl time.Duration, c Credentials) (time.Duration, error) {
	if ttl < time.Second || ttl > time.Hour {
		return 0, fail("INVALID_INPUT")
	}
	if !c.Expiration.IsZero() {
		remaining := time.Until(c.Expiration) - 10*time.Second
		if remaining < time.Second {
			return 0, fail("CONFIGURATION")
		}
		if ttl > remaining {
			ttl = remaining
		}
	}
	return ttl, nil
}
func statusError(status int) error {
	switch status {
	case 404:
		return fail("NOT_FOUND")
	case 401, 403:
		return fail("FORBIDDEN")
	case 409, 412:
		return fail("CONFLICT")
	}
	return &Error{Code: "UNAVAILABLE", Retryable: status == 429 || status >= 500}
}

type Cloud struct {
	name   string
	sign   func(context.Context, string, string, int64, string, time.Duration, bool) (SignedRequest, error)
	head   func(context.Context, string) (ObjectInfo, error)
	remove func(context.Context, string) error
	list   func(context.Context, ListInput) (Page, error)
}

func (p *Cloud) Name() string { return p.name }
func (p *Cloud) Capabilities() Capabilities {
	return Capabilities{DirectUpload: true, Extensions: []string{}}
}
func (p *Cloud) SignUpload(ctx context.Context, key string, size int64, contentType string, ttl time.Duration) (SignedRequest, error) {
	if e := ValidKey(key); e != nil {
		return SignedRequest{}, e
	}
	if e := validInput(PutInput{Body: strings.NewReader(""), Size: size, ContentType: contentType}); e != nil {
		return SignedRequest{}, e
	}
	return p.sign(ctx, "PUT", key, size, contentType, ttl, true)
}
func (p *Cloud) SignDownload(ctx context.Context, key string, ttl time.Duration) (SignedRequest, error) {
	if e := ValidKey(key); e != nil {
		return SignedRequest{}, e
	}
	return p.sign(ctx, "GET", key, 0, "", ttl, true)
}
func (p *Cloud) Head(ctx context.Context, key string) (ObjectInfo, error) {
	if e := ValidKey(key); e != nil {
		return ObjectInfo{}, e
	}
	ctx, cancel := context.WithTimeout(ctx, time.Minute)
	defer cancel()
	v, e := p.head(ctx, key)
	return v, Normalize(e)
}
func (p *Cloud) Delete(ctx context.Context, key string) error {
	if e := ValidKey(key); e != nil {
		return e
	}
	ctx, cancel := context.WithTimeout(ctx, time.Minute)
	defer cancel()
	e := p.remove(ctx, key)
	if IsCode(e, "NOT_FOUND") {
		return nil
	}
	return Normalize(e)
}
func (p *Cloud) List(ctx context.Context, input ListInput) (Page, error) {
	n, e := limit(input.Limit)
	if e != nil {
		return Page{}, e
	}
	if len(input.Cursor) > 8192 || len(input.Prefix) > 1024 {
		return Page{}, fail("INVALID_INPUT")
	}
	input.Limit = n
	ctx, cancel := context.WithTimeout(ctx, time.Minute)
	defer cancel()
	v, e := p.list(ctx, input)
	return v, Normalize(e)
}
func (p *Cloud) Get(ctx context.Context, key string) (Download, error) {
	if e := ValidKey(key); e != nil {
		return Download{}, e
	}
	s, e := p.sign(ctx, "GET", key, 0, "", 5*time.Minute, false)
	if e != nil {
		return Download{}, e
	}
	req, e := http.NewRequestWithContext(ctx, "GET", s.URL, nil)
	if e != nil {
		return Download{}, Normalize(e)
	}
	for k, v := range s.Headers {
		req.Header.Set(k, v)
	}
	res, e := (&http.Client{Timeout: 5 * time.Minute}).Do(req)
	if e != nil {
		return Download{}, Normalize(e)
	}
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		res.Body.Close()
		return Download{}, statusError(res.StatusCode)
	}
	if res.ContentLength < 0 {
		res.Body.Close()
		return Download{}, fail("UNAVAILABLE")
	}
	return Download{Body: res.Body, Info: ObjectInfo{Key: key, Size: res.ContentLength, ContentType: res.Header.Get("Content-Type"), ETag: res.Header.Get("ETag"), ModifiedAt: res.Header.Get("Last-Modified")}}, nil
}

type exactReader struct {
	r          io.Reader
	size, seen int64
}

func (r *exactReader) Read(b []byte) (int, error) {
	n, e := r.r.Read(b)
	r.seen += int64(n)
	if r.seen > r.size || (e == io.EOF && r.seen != r.size) {
		return n, fail("INVALID_INPUT")
	}
	return n, e
}
func (p *Cloud) Put(ctx context.Context, key string, input PutInput) (ObjectInfo, error) {
	if e := ValidKey(key); e != nil {
		return ObjectInfo{}, e
	}
	if e := validInput(input); e != nil {
		return ObjectInfo{}, e
	}
	s, e := p.sign(ctx, "PUT", key, input.Size, input.ContentType, 5*time.Minute, false)
	if e != nil {
		return ObjectInfo{}, e
	}
	body := &exactReader{r: input.Body, size: input.Size}
	req, e := http.NewRequestWithContext(ctx, "PUT", s.URL, body)
	if e != nil {
		return ObjectInfo{}, Normalize(e)
	}
	req.ContentLength = input.Size
	if input.Size == 0 {
		var b [1]byte
		n, e := body.Read(b[:])
		if n != 0 || e != io.EOF {
			return ObjectInfo{}, fail("INVALID_INPUT")
		}
		req.Body = http.NoBody
	}
	for k, v := range s.Headers {
		req.Header.Set(k, v)
	}
	req.Header.Set("Content-Type", input.ContentType)
	res, e := (&http.Client{Timeout: 5 * time.Minute}).Do(req)
	if e != nil {
		return ObjectInfo{}, Normalize(e)
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return ObjectInfo{}, statusError(res.StatusCode)
	}
	if body.seen != input.Size {
		return ObjectInfo{}, fail("INVALID_INPUT")
	}
	return ObjectInfo{Key: key, Size: input.Size, ContentType: input.ContentType, ETag: res.Header.Get("ETag")}, nil
}
